### PR TITLE
[NT-2120] Migrating query for fetching user backings to Apollo.

### PR DESCRIPTION
**Pull request by singhhari**
# 📲 What

One of the ways we fetch a users backings is by using a GraphQL query that provides us all  backings for a given user. The query technically works for all backings, regardless of `BackingState`  but our specific use case on this application is just for errored backings. This PR will handle the migration from our legacy DSL to Apollo GraphQL client. 

# 🤔 Why

This is simply another step forward in our larger implementation of Apollo GraphQL. We are migrating all queries, migrations and networking logic to Apollo for simplicity of use.

# 🛠 How

- Rename `BackingsEnvelope` to `ErroredBackingsEnvelope` to more accurately reflect the data types usage.
- Update the `fetchGraphUserBackings` method in our `Service` object to use Apollo GraphQL client.
- Add a new query for fetching backings.
- Create an adapter that will create a `SignalProducer` from the `Data` type created in our `GraphAPI` class.

Note: Testing this required me to pair with a web engineer to manually trigger a failed payment error on staging. The testing was successful and I saw the Activity tab update with a failed error update. This will be difficult to recreate for the reviewer of the PR unless following the same steps.

# ✅ Acceptance criteria

- [ ] Launch the app on staging and verify the query to fetch user backings is occurring. It usually occurs a second or two after the tab bar renders. You can use Charles to observe the very first graph request.
- [ ] Verify the query is looking for errored backings and there is no regression
- [ ] Optional: Pair with a payments engineer to go through the steps outlined here: 
https://app.getguru.com/card/cn9ELeri/How-to-test-errored-pledge-flow-in-Kickstarter-in-staging

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: KsApi/models/ErroredBackingsEnvelope.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+extension ErroredBackingsEnvelope {
+  internal static func envelopeProducer(
+    from envelope: UserEnvelope<GraphBackingEnvelope>
+  ) -> SignalProducer<ErroredBackingsEnvelope, ErrorEnvelope> {
+    let envelopes = envelope.me.backings.nodes.compactMap { graphBacking -> ProjectAndBackingEnvelope? in
+      guard
+        let backing = Backing.backing(from: graphBacking),
+        let graphProject = graphBacking.project,
+        let project = Project.project(from: graphProject)
```
Comment (singhhari) [inline KsApi/models/ErroredBackingsEnvelope.swift:18]: I did not change this method but I think reformatting flagged codecov.

File: KsApi/models/ErroredBackingsEnvelope.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+   - parameter from: The `GraphAPI.FetchUserBackingsQuery.Data` object that contains the backings and fragment details.
+   */
+  static func producer(from data: GraphAPI.FetchUserBackingsQuery
+    .Data) -> SignalProducer<ErroredBackingsEnvelope, ErrorEnvelope> {
+    guard let envelopes = data.me?.backings?.nodes?.compactMap({ backing -> ProjectAndBackingEnvelope? in
+      guard let backingFragment = backing?.fragments.backingFragment,
+        let backing = Backing.backing(from: backingFragment),
+        let projectFragment = backingFragment.project?.fragments.projectFragment,
+        let project = Project.project(from: projectFragment)
```
Comment (singhhari) [inline KsApi/models/ErroredBackingsEnvelope.swift:38]: Can't test these producers until we get a mock GraphQL client made. We should prioritize that for next sprint and come back to fix all of these untested adapter methods.

**Review by msadoon — CHANGES_REQUESTED** [review]
Some changes (discussed in 1-1 as well).

File: KsApi/models/ErroredBackingsEnvelope.swift (hunk 1 above)
Comment (singhhari) [reply KsApi/models/ErroredBackingsEnvelope.swift:18]: Deleted in the last commit.

File: KsApi/models/ErroredBackingsEnvelope.swift (hunk 2 above)
Comment (singhhari) [reply KsApi/models/ErroredBackingsEnvelope.swift:38]: Added a test in the last commit that should hold us over until we add the mock client.

File: KsApi/Service.swift [hunk 3]
```diff
@@ -296,11 +296,14 @@ public struct Service: ServiceType {
       .flatMap(UserEnvelope<GraphUser>.envelopeProducer(from:))
   }
 
-  public func fetchGraphUserBackings(query: NonEmptySet<Query>)
-    -> SignalProducer<BackingsEnvelope, ErrorEnvelope> {
-    return fetch(query: query)
-      .mapError(ErrorEnvelope.envelope(from:))
-      .flatMap(BackingsEnvelope.envelopeProducer(from:))
+  public func fetchGraphUserBackings(status: BackingState)
+    -> SignalProducer<ErroredBackingsEnvelope, ErrorEnvelope> {
```
[the remark this answers is not in the document]
Comment (singhhari) [reply KsApi/Service.swift:300]: Renamed method to `fetchErroredUserBackings`.

File: KsApi/queries/FetchUserBackings.graphql [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            }
+            limit
+            limitPerBacker
+            name
+            remainingQuantity
+            shippingPreference
+            shippingRules {
+              ...ShippingRuleFragment
+            }
+            startsAt
```
[the remark this answers is not in the document]
Comment (singhhari) [reply KsApi/queries/FetchUserBackings.graphql:34]: good catch!

**Review by msadoon — APPROVED** [review]
yep, good to go!

**Pull request merged**